packages/server/lib/util/resourceTypeAndCredentialManager.ts

```diff
@@ -0,0 +1,88 @@
+import md5 from 'md5'
+import Debug from 'debug'
+import type { RequestCredentialLevel, RequestResourceType } from '@packages/proxy'
+
+type AppliedCredentialByUrlAndResourceMap = Map<string, Array<{
+  resourceType: RequestResourceType
+  credentialStatus: RequestCredentialLevel
+}>>
+
+const debug = Debug('cypress:server:util:resource-type-and-credential')
+
+const hashUrl = (url: string): string => {
+  return md5(decodeURIComponent(url))
+}
+
+// leverage a singleton Map throughout the server to prevent clashes with this context bindings
+const _appliedCredentialByUrlAndResourceMap: AppliedCredentialByUrlAndResourceMap = new Map()
+
+class ResourceTypeAndCredentialManagerClass {
+  get (url: string, optionalResourceType?: RequestResourceType): {
+    resourceType: RequestResourceType
+    credentialStatus: RequestCredentialLevel
+  } {
+    const hashKey = hashUrl(url)
+
+    debug(`credentials request received for request url ${url}, hashKey ${hashKey}`)
+    let value: {
+      resourceType: RequestResourceType
+      credentialStatus: RequestCredentialLevel
+    } | undefined
+
+    const credentialsObj = _appliedCredentialByUrlAndResourceMap.get(hashKey)
+
+    if (credentialsObj) {
+      // remove item from queue
+      value = credentialsObj?.shift()
+      debug(`credential value found ${value}`)
+    }
+
+    // if value is undefined for any reason, apply defaults and assume xhr if no optionalResourceType
+    // optionalResourceType should be provided with CDP based browsers, so at least we have a fallback that is more accurate
+    if (value === undefined) {
+      value = {
+        resourceType: optionalResourceType || 'xhr',
+        credentialStatus: optionalResourceType === 'fetch' ? 'same-origin' : false,
+      }
+    }
+
+    return value
+  }
+
+  set ({ url,
+    resourceType,
+    credentialStatus,
+  }: {
+    url: string
+    resourceType: RequestResourceType
+    credentialStatus: RequestCredentialLevel
+  }) {
+    const hashKey = hashUrl(url)
+
+    debug(`credentials request stored for request url ${url}, resourceType ${resourceType}, hashKey ${hashKey}`)
+
+    let urlHashValue = _appliedCredentialByUrlAndResourceMap.get(hashKey)
+
+    if (!urlHashValue) {
+      _appliedCredentialByUrlAndResourceMap.set(hashKey, [{
+        resourceType,
+        credentialStatus,
+      }])
+    } else {
+      urlHashValue.push({
+        resourceType,
+        credentialStatus,
+      })
+    }
+  }
+
+  clear () {
+    _appliedCredentialByUrlAndResourceMap.clear()
+  }
+}
+
+// export as a singleton
+export const resourceTypeAndCredentialManager = new ResourceTypeAndCredentialManagerClass()
+
+// export but only as a type
+export type ResourceTypeAndCredentialManager = ResourceTypeAndCredentialManagerClass
```

Comment:
You can export the class and the importer can do `import type` to just get the type, right?

Reply:
Yep thats a way to do it. I explicitly exported the type here because I do not want anyone instantiating the `ResourceTypeAndCredentialManagerClass` since it is supposed to be a singleton, but the constructor should likely just create multiple instances of the class that ultimately point to the same map.